quickwit/quickwit-indexing/src/models/split_fields.rs

```diff
@@ -0,0 +1,202 @@
+    pub indexed: bool,
+    pub stored: bool,
+    pub fast: bool,
+}
+
+impl FieldConfig {
+    fn serialize(&self) -> [u8; 2] {
+        let typ = self.typ.to_code();
+        let flags = (self.indexed as u8) << 2 | (self.stored as u8) << 1 | (self.fast as u8);
+        [typ, flags]
+    }
+    fn deserialize_from(data: [u8; 2]) -> io::Result<FieldConfig> {
+        let typ = Type::from_code(data[0]).ok_or_else(|| {
+            io::Error::new(
+                ErrorKind::InvalidData,
+                format!("could not deserialize type {}", data[0]),
+            )
+        })?;
+
+        let data = data[1];
+        let indexed = (data & 0b100) != 0;
+        let stored = (data & 0b010) != 0;
+        let fast = (data & 0b001) != 0;
+
+        Ok(FieldConfig {
+            typ,
+            indexed,
+            stored,
+            fast,
+        })
+    }
+}
+
+/// Serializes the Split fields.
+///
+/// `fields_metadata` has to be sorted.
+pub fn serialize_split_fields(fields_metadata: &[FieldMetadata]) -> Vec<u8> {
+    debug_assert!(fields_metadata.windows(2).all(|w| w[0] < w[1]));
+    let mut out = Vec::new();
+    // Write Num Fields
+    let length = fields_metadata.len() as u32;
+    out.extend_from_slice(&length.to_le_bytes());
+
+    for field_metadata in fields_metadata {
+        let field_config = FieldConfig {
+            typ: field_metadata.typ,
+            indexed: field_metadata.indexed,
+            stored: field_metadata.stored,
+            fast: field_metadata.fast,
+        };
+
+        // Write Config 2 bytes
+        out.extend_from_slice(&field_config.serialize());
+        let str_length = field_metadata.field_name.len() as u16;
+        // Write String length 2 bytes
+        out.extend_from_slice(&str_length.to_le_bytes());
+        out.extend_from_slice(field_metadata.field_name.as_bytes());
+    }
+    zstd::stream::encode_all(&mut &out[..], 3).expect("zstd encoding failed")
+}
+
+/// Reads a fixed number of bytes into an array and returns the array.
+fn read_exact_array<R: Read, const N: usize>(reader: &mut R) -> io::Result<[u8; N]> {
+    let mut buffer = [0u8; N];
+    reader.read_exact(&mut buffer)?;
+    Ok(buffer)
+}
+
+/// Reads the Split fields from a zstd compressed stream of bytes
+pub fn iter_split_fields<R: Read>(
+    reader: R,
+) -> io::Result<impl Iterator<Item = io::Result<FieldMetadata>>> {
+    let reader = zstd::Decoder::new(reader)?;
+    _iter_split_fields(reader)
+}
+
+/// Reads the Split fields from a stream of bytes
+fn _iter_split_fields<R: Read>(
+    mut reader: R,
+) -> io::Result<impl Iterator<Item = io::Result<FieldMetadata>>> {
+    let mut num_fields = u32::from_le_bytes(read_exact_array::<_, 4>(&mut reader)?);
+
+    Ok(std::iter::from_fn(move || {
+        if num_fields == 0 {
+            return None;
+        }
+        num_fields -= 1;
+
+        // Helper closure to use `?` for error handling
+        let read_field = |reader: &mut R| -> io::Result<FieldMetadata> {
```

Comment:
can we make it a function?

Reply:
that's typically the kind of function you want to unit test.

Reply:
done